Detectors/TPC/calibration/run/calib-pedestal.cxx

```diff
@@ -47,53 +63,167 @@ void printRDH(const RDH& rdh)
-                    << dh->subSpecification << " payload size " << dh->payloadSize;
-
-          // there is a bug in InpuRecord::get for vectors of simple types, not catched in
-          // DataAllocator unit test
-          //auto data = inputs.get<std::vector<char>>(input.spec->binding.c_str());
-          //LOG(INFO) << "data size " << data.size();
-          printHeader();
-          try {
-            const char* pos = input.payload;
-            const char* last = pos + dh->payloadSize;
-
-            o2::framework::RawParser parser(input.payload, dh->payloadSize);
-
-            //while (pos < last) {
-            for (auto it = parser.begin(), end = parser.end(); it != end; ++it) {
-              auto* rdhPtr = it.get_if<o2::header::RAWDataHeaderV4>();
-              if (!rdhPtr) {
+  // set up configuration
+  o2::conf::ConfigurableParam::updateFromFile(config.options().get<std::string>("configFile"));
+  o2::conf::ConfigurableParam::updateFromString(config.options().get<std::string>("configKeyValues"));
+  o2::conf::ConfigurableParam::writeINI("o2tpccalibration_configuration.ini");
+
+  struct ProcessAttributes {
+    CalibPedestal calibPedestal;
+    rawreader::RawReaderCRUManager rawReader;
+    uint32_t lastOrbit{0};
+    uint64_t lastTFID{0};
+    uint32_t maxEvents{100};
+    bool quit{false};
+    bool dumped{false};
+  };
+
+  auto initFunction = [](InitContext& ic) {
+    auto processAttributes = std::make_shared<ProcessAttributes>();
+    // set up calibration
+    // TODO:
+    // it is a bit ugly to use the RawReaderCRUManager for this is.
+    // At some point the raw reader code should be cleaned up and modularized
+    {
+      auto& pedestal = processAttributes->calibPedestal;
+      processAttributes->rawReader.createReader("");
+      processAttributes->rawReader.setADCDataCallback([&pedestal](const PadROCPos& padROCPos, const CRU& cru, const gsl::span<const uint32_t> data) -> Int_t {
+        Int_t timeBins = pedestal.update(padROCPos, cru, data);
+        pedestal.setNumberOfProcessedTimeBins(std::max(pedestal.getNumberOfProcessedTimeBins(), size_t(timeBins)));
+        return timeBins;
+      });
+      processAttributes->maxEvents = static_cast<uint32_t>(ic.options().get<int>("max-events"));
+    }
+
+    auto processingFct = [processAttributes](ProcessingContext& pc) {
+      // in case the maximum number of events was reached don't do further processing
+      if (processAttributes->quit) {
+        return;
+      }
+
+      //if (pc.inputs().isValid("TFID")) {
+        //auto tfid = pc.inputs().get<uint64_t>("TFID");
+        //LOG(INFO) << "TFid: " << tfid;
+        //processAttributes->lastTFID = tfid;
+      //}
+
+      for (auto& input : pc.inputs()) {
+        const auto* dh = DataRefUtils::getHeader<o2::header::DataHeader*>(input);
+
+        // ---| only process RAWDATA, there might be a nicer way to do this |---
+        if (dh->dataDescription != o2::header::gDataDescriptionRawData) {
```

Comment:
```suggestion
        if (dh == nullptr || dh->dataDescription != o2::header::gDataDescriptionRawData) {
```

Reply:
When using custom completion policies, input slots might be invalid, the `DataRef` object contains `nullptr`.

I'm going to enhance the API in order to simplify implementations and to make it more intuitive. Actually the API has not been adopted to features which have been added.

Reply:
#2843 will silently fix the problem, invalid slots will be filtered out from the iteration

Reply:
Thanks for the quick improvement. Meanwhile I also tested with the `nullptr`. This works. Thanks for the hint. Next I'll try to use TFID for event detection.

What would even be better is to get all data belonging to one TF in one call, instead of one call per link.

Reply:
Thanks for the quick improvement. Meanwhile I also tested with the `nullptr`. This works. Thanks for the hint. Next I'll try to use TFID for event detection.

What would even be better is to get all data belonging to one TF in one call, instead of one call per link.